pkg/apis/networking/validation/validation.go

```diff
@@ -204,13 +207,17 @@ func validateIngressTLS(spec *networking.IngressSpec, fldPath *field.Path) field
 }
 
 // ValidateIngressSpec tests if required fields in the IngressSpec are set.
-func ValidateIngressSpec(spec *networking.IngressSpec, fldPath *field.Path) field.ErrorList {
+func ValidateIngressSpec(spec *networking.IngressSpec, fldPath *field.Path, requestGV schema.GroupVersion) field.ErrorList {
 	allErrs := field.ErrorList{}
-	// TODO: Is a default backend mandatory?
-	if spec.Backend != nil {
-		allErrs = append(allErrs, validateIngressBackend(spec.Backend, fldPath.Child("backend"))...)
+	if spec.DefaultBackend != nil {
+		allErrs = append(allErrs, validateIngressBackend(spec.DefaultBackend, fldPath.Child("defaultBackend"))...)
 	} else if len(spec.Rules) == 0 {
-		allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, "either `backend` or `rules` must be specified"))
+		if requestGV == networkingv1.SchemeGroupVersion {
+			allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, "either `defaultBackend` or `rules` must be specified"))
+		}
+		if requestGV == networkingv1beta1.SchemeGroupVersion {
+			allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, "either `backend` or `rules` must be specified"))
+		}
```

Comment:
this is fragile, and means if we add another version we have to remember to revisit this or we get unvalidated fields!

assuming we are better at naming this field now than we were four years ago, something like this that assumes future versions will continue using the v1 field name might make more sense:
```go
var defaultBackendFieldName string
switch requestGV {
case networkingv1beta1.SchemeGroupVersion, extensionsv1beta1.SchemeGroupVersion:
	defaultBackendFieldName = "backend"
default:
	defaultBackendFieldName = "defaultBackend"
}

allErrs := field.ErrorList{}
if spec.DefaultBackend != nil {
	allErrs = append(allErrs, validateIngressBackend(spec.DefaultBackend, fldPath.Child(defaultBackendFieldName))...)
} else if len(spec.Rules) == 0 {
	allErrs = append(allErrs, field.Invalid(fldPath, spec.Rules, fmt.Sprintf("either `%s` or `rules` must be specified", defaultBackendFieldName)))
}
```

Reply:
When I tried it this way it doesn't seem to trigger the validation error that is expected in strategy_test.go. I'm scouring the relevant files under pkg/registry/networking/ingress/* to see if I am accidentally using the old backend type but I have not found a mistake yet.

Reply:
Nvm copy/paste error, I found my mistake

Reply:
Actually even after fixing an indenting problem, this still doesnt work because it passes only a string to the ingressbackend validation. I kept the if/else for now.

Reply:
also, this code doesn't currently handle extensions/v1beta1